Add support for WordPress 6.1

zaerl opened the pull request:
This PR upgrades the plugin to 0.3 and adds these fixes:

1. Tested with WordPress `6.1`
2. Tested with PHP `7.4.33`
3. Remove the deprecated `screen_icon` function for WordPress >= 3.8.0
4. Substitute deprecated `$wpdb->escape` function
5. Additional check on an empty list of categories
6. Additional check of the number of links inserted
7. Check if the link manager (needed by this plugin) is enabled for the WordPress > 3.5 installations where it is [hidden by default](https://wordpress.org/support/wordpress-version/version-3-5/#links). This mimics the behavior found in [wp_link_manager_disabled_message](https://developer.wordpress.org/reference/functions/wp_link_manager_disabled_message/).

How to test:
1. Clone the plugin
2. Spin a new docker image[^1]
3. Substitute `YOUR_FOLDER` with the path of the cloned plugin
4. Go to `http://localhost/wp-admin/plugins.php` and activate **OPML Importer**
5. Open `http://localhost/wp-admin/admin.php?import=opml`
6. It will ask you to enable the link manager if you don't have it
7. Import an OPML file
9. The plugin must load correctly, and no errors displayed

[^1]: `docker-compose.yml` file
```yaml
services:
  db:
    image: mariadb:latest
    command: '--default-authentication-plugin=mysql_native_password'
    volumes:
      - db_data:/var/lib/mysql
    restart: always
    environment:
      - MYSQL_ROOT_PASSWORD=somewordpress
      - MYSQL_DATABASE=wordpress
      - MYSQL_USER=wordpress
      - MYSQL_PASSWORD=wordpress
    expose:
      - 3306
      - 33060
  wordpress:
    image: wordpress:latest
    volumes:
      - wp_data:/var/www/html
      - YOUR_FOLDER:/var/www/html/wp-content/plugins/opml-importer
    ports:
      - 80:80
    restart: always
    environment:
      - WORDPRESS_DB_HOST=db
      - WORDPRESS_DB_USER=wordpress
      - WORDPRESS_DB_PASSWORD=wordpress
      - WORDPRESS_DB_NAME=wordpress
      - WORDPRESS_DEBUG=1
volumes:
  db_data:
  wp_data:
```

zaerl committed:
Fix: remove deprecated function

zaerl committed:
Automatically enable the link manager if the plugin is enabled.

zaerl committed:
Fix: add check for missing categories.

zaerl committed:
Fix: substitute deprecated function.

zaerl committed:
Fix: wrong calculation of inserted links.

jrfnl commented on opml-importer.php:
```diff
@@ -10,6 +10,9 @@
 License: GPL version 2 or later - http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 */
 
+// If a user want to import links we must enable the link manager
+add_filter( 'pre_option_link_manager_enabled', '__return_true' );
```
This will only have effect while the importer is installed and active. As importers are - in most cases - used once and then deactivated again, I don't think this is very effective.

jrfnl commented on opml-importer.php:
```diff
@@ -50,7 +53,6 @@ function dispatch() {
 ?>
 
 <div class="wrap">
-<?php screen_icon(); ?>
```
This plugin [still supports WP 3.0](https://github.com/WordPress/opml-importer/blob/1c9caeeb2116d1bff4d08eaf07ad280644b1d552/readme.txt#L5) - should this function maybe be called conditionally ?

zaerl replied:
Called only when WordPress is < 3.8.0.

jrfnl commented on opml-importer.php:
```diff
@@ -71,15 +73,20 @@ function dispatch() {
+<?php
+$categories = get_terms('link_category', array('get' => 'all'));
+
+if ( count($categories) ) {
```
If you're going to add defensive coding...
* Should a check for `WP_Error` being returned be added ?
* `get_terms()` does not always return an array, so `count()` on its own is not sufficient. Maybe check with `is_array() && ! empty()` ?

Ref: https://developer.wordpress.org/reference/functions/get_terms/

jrfnl commented on opml-importer.php:
```diff
@@ -71,15 +73,20 @@ function dispatch() {
 <?php _e('Category:', 'opml-importer') ?> <select name="cat_id" id="cat_id">
 <?php
-$categories = get_terms('link_category', array('get' => 'all'));
 foreach ($categories as $category) {
 ?>
 <option value="<?php echo $category->term_id; ?>"><?php echo esc_html(apply_filters('link_category', $category->name)); ?></option>
 <?php
 } // end foreach
```
Should this foreach be indented now it is nested in the `if` ?

jrfnl commented on opml-importer.php:
```diff
@@ -102,9 +109,13 @@ function dispatch() {
-		if ( $cat_id < 1 )
-			$cat_id  = 1;
+		$cat_id = null;
+		if( array_key_exists( 'cat_id', $_POST ) ) {
```
```suggestion
		if( isset( $_POST['cat_id'] ) ) {
```

jrfnl commented on opml-importer.php:
```diff
@@ -102,9 +109,13 @@ function dispatch() {
+		if( array_key_exists( 'cat_id', $_POST ) ) {
+			$cat_id = abs( (int) $_POST['cat_id'] );
+
+			if ( $cat_id < 1 )
+				$cat_id  = 1;
```
Please use braces for control structures.

jrfnl commented on opml-importer.php:
```diff
@@ -102,9 +109,13 @@ function dispatch() {
-		$cat_id = abs( (int) $_POST['cat_id'] );
-		if ( $cat_id < 1 )
-			$cat_id  = 1;
+		$cat_id = null;
```
To maintain the same behaviour as before (minus the potential "undefined index" notice), this should be `1`, not `null`. See: https://3v4l.org/XfCju

This will also affect the rest of the code below which now contains the presumption of `null` having special menaing.

zaerl replied:
Thanks for this review. I'm fixing everything. Here I've used `null` because in the old implementation, a `'link_category' => 1` was passed to `wp_insert_link` when the user didn't have any categories (and so didn't choose something). The ``wp_insert_link`` function states that `If 'link_category' is empty, uses default link category.`

So I thought it was the best thing to do instead of passing `1`. I considered passing `1` as a bug. What do you think about this?

jrfnl replied:
For a new site, passing `'link_category' => 1` effectively is the _same_ as "uses default link category". For a pre-existing site for which new information is being added if _may_ be different, so yes, it could be considered a bug fix, but it is also a change in the behaviour of the plugin, which _may_ or _may not_ be desired.

Your PR description doesn't mention anything about changing the behaviour of the plugin, so I reviewed based on it purely being a update for WP compatibility and defensive coding and flagged this change because it doesn't belong in those categories.

I suggest maintaining the same behaviour as before for this PR and submitting a separate PR proposing the behavioural change as that's a different (merge-)decision point than for this PR.

zaerl replied:
Ok, I agree.

jrfnl commented on opml-importer.php:
```diff
@@ -133,20 +144,26 @@ function dispatch() {
-				$link = array( 'link_url' => $urls[$i], 'link_name' => $wpdb->escape($names[$i]), 'link_category' => array($cat_id), 'link_description' => $wpdb->escape($descriptions[$i]), 'link_owner' => $user_ID, 'link_rss' => $feeds[$i]);
-				wp_insert_link($link);
-				echo sprintf('<p>'.__('Inserted <strong>%s</strong>', 'opml-importer').'</p>', $names[$i]);
+				$link = array( 'link_url' => $urls[$i], 'link_name' => esc_sql($names[$i]), 'link_category' => is_null($cat_id) ? null : array($cat_id), 'link_description' => esc_sql($descriptions[$i]), 'link_owner' => $user_ID, 'link_rss' => $feeds[$i]);
```
Not sure `esc_sql()` is needed - [`wp_insert_link()`](https://developer.wordpress.org/reference/functions/wp_insert_link/) claims to sanitize all data it receives.

jrfnl reviewed (changes requested):
Looks like this needs more work.

zaerl committed:
Add support for WordPress < 3.8.0

zaerl committed:
Various fixes.

https://github.com/WordPress/opml-importer/pull/2#discussion_r1022393712
https://github.com/WordPress/opml-importer/pull/2#discussion_r1022396419
https://github.com/WordPress/opml-importer/pull/2#discussion_r1022397818
https://github.com/WordPress/opml-importer/pull/2#discussion_r1022400432
https://github.com/WordPress/opml-importer/pull/2#discussion_r1022404953
https://github.com/WordPress/opml-importer/pull/2#discussion_r1022446685

zaerl committed:
Fix: change the category default value.

zaerl committed:
Fix: remove forced Link Manager filter

zaerl commented:
@jrfnl I've applied new changes, and I hope these will cover all the needed changes. Can you please do another review? Thanks.

jrfnl commented:
> @jrfnl I've applied new changes, and I hope these will cover all the needed changes. Can you please do another review? Thanks.

I'm not the maintainer, I was just passing by and did a quick review.

zaerl commented:
> > @jrfnl I've applied new changes, and I hope these will cover all the needed changes. Can you please do another review? Thanks.
> 
> I'm not the maintainer, I was just passing by and did a quick review.

Ok, thanks anyway. 🙂

zaerl committed:
Fix: code style.

zaerl committed:
Version upgrade.

vishnugopal commented:
I noticed that when I go to install this plugin, it asks for installation of [Link Manager](https://wordpress.org/plugins/link-manager/). We probably need to support Link Manager too, and keep it updated @zaerl

vishnugopal reviewed (approved):
Works fine! Thanks for these changes!
